stairway/src/main/java/bio/terra/stairway/FlightContext.java

```diff
@@ -48,4 +49,23 @@ public interface FlightContext {
 
   /** @return pretty string describing the flight */
   String flightDesc();
+
+  /**
+   * Return meter data given the meter name
+   *
+   * @param name name of the meter to lookup
+   * @return optional progress meter data
+   */
+  Optional<ProgressMeterData> getProgressMeter(String name);
```

Comment:
What's the benefit of returning `Optional` here? The only case where returning `Optional` is needed is if you need to distinguish between a `null` value and a missing value. Is there a case where a `null` result is different from a missing result?

If a caller wants to use `Optional` to examine this result, they can always use `Optional.ofNullable(getProgressMeter(...))`.

Reply:
Hmm... that is not my understanding of the only case for returning Optional. I'm still titrating on its use. I thought that returning Optional was preferred to returning null since it forces the caller to code for a missing return.

But I don't feel strongly about it. I ain't afraid of no nulls!

Reply:
@jmczerk do you have an opinion?

It would probably be more consistent with the rest of the interface to remove the Optional.

Reply:
No strong opinion here.  If you expect people to generally ask for progress meters that exist, I'd say keep it bare in favor of consistency with the rest of the interface.  If you expect consuming code to be asking for meters speculatively, then maybe `Optional` would make more sense?

Reply:
> Keep the Optional. I never want to see a method return null if it can be helped.

Can you provide a reasoning behind this opinion? My original question still remains: what is the benefit of returning `Optional` over `null`? If the answer is "I don't like `null`", that's not very compelling, and is not aligned with with Java implementor's intention of this language feature.

Reply:
Before there was Optional, callers had to remember or guess or learn from experience when a function could return null, and implement ugly logic to handle it. Now we have a self documenting, functional construct that supports transformations and explicit presence and absence checks. I think Enterprise Java 3rd edition had a good chapter on it. It's a higher level of abstraction than null too, which is a good thing. I think it's also fair to say we've reduced the occurrence of NPEs in WSM significantly by using it brutally.

Reply:
> Before there was Optional, callers had to remember or guess or learn from experience when a function could return null

First, `Optional` itself can be `null` -- just by declaring that a function returns `Optional` doesn't mean that someone couldn't return a null value from it. In this case we are writing code for our own consumption, so I think we can do better than guesswork and experimentation to determine how an API works. I would recommend good documentation, and a convention of never returning `null` from a method that returns a collection.

> and implement ugly logic to handle it.

Ugliness is a subjective measure. Functional forms are nice but don't always make for (what I would consider to be) idiomatic Java. Writing Java that's trying to be Scala looks ugly to me.

> It's a higher level of abstraction than null too, which is a good thing.

I disagree with this; I would characterize is as an unnecessary wrapper around objects which, except for the case where one needs to distinguish missing from `null`, adds no value. It sounds like we've had different experiences with coding however:

> I think it's also fair to say we've reduced the occurrence of NPEs in WSM significantly by using it brutally.

Anecdotal data is hard to prove or disprove; all I can say is personally I've worked without using `Optional` for a long time and haven't seen a need for this feature. I've also worked in languages that protect against nulls, like Scala, and have had NPE issues. (I've also used languages that natively support non-nullable types, those are great but Java doesn't have this.)

If we were working in a language that supported non-null types (like C#, Kotlin, etc) or idiomatic null handling (like Scala) then this would make sense. But this is Java and (again IMO) such forms are not idiomatic to Java.

Reply:
I removed the Optionals. It is more consistent with the bulk of the Stairway API.